src/bulkUtils.ts

```diff
@@ -5,49 +5,53 @@
  * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
  */
 
-import { IngestJobV2, IngestJobV2Results, IngestOperation, JobInfoV2 } from 'jsforce/lib/api/bulk.js';
-import { Schema } from 'jsforce';
+import type {
+  JobInfoV2,
+  IngestJobV2Results,
+  IngestJobV2SuccessfulResults,
+  IngestJobV2FailedResults,
+  IngestJobV2UnprocessedRecords,
+} from '@jsforce/jsforce-node/lib/api/bulk2.js';
+
+import { Schema } from '@jsforce/jsforce-node';
 import { Connection, Messages } from '@salesforce/core';
 import { BulkProcessedRecordV2, BulkRecordsV2 } from './types.js';
 
 Messages.importMessagesDirectoryFromMetaUrl(import.meta.url);
 const messages = Messages.loadMessages('@salesforce/plugin-data', 'messages');
 
-const POLL_FREQUENCY_MS = 5000;
+export const POLL_FREQUENCY_MS = 5000;
 
 export const isBulkV2RequestDone = (jobInfo: JobInfoV2): boolean =>
   ['Aborted', 'Failed', 'JobComplete'].includes(jobInfo.state);
 
 export const transformResults = (results: IngestJobV2Results<Schema>): BulkRecordsV2 => ({
-  successfulResults: results.successfulResults.map((record) => record as unknown as BulkProcessedRecordV2),
-  failedResults: results.failedResults.map((record) => record as unknown as BulkProcessedRecordV2),
-  unprocessedRecords: results.unprocessedRecords.map((record) => record as unknown as BulkProcessedRecordV2),
+  // ensureArray is used to handle the undefined or non-array case
+  successfulResults: results.successfulResults.map(anyRecordToBulkProcessedRecordV2),
+  failedResults: results.failedResults.map(anyRecordToBulkProcessedRecordV2),
+  // if the csv can't be read, it returns a string that is the csv body
+  ...(typeof results.unprocessedRecords === 'string'
+    ? { unprocessedRecords: [], unparsed: results.unprocessedRecords }
+    : { unprocessedRecords: results.unprocessedRecords.map(anyRecordToBulkProcessedRecordV2) }),
 });
 
-export const validateSobjectType = async (sobjectType: string, connection: Connection): Promise<void> => {
+const anyRecordToBulkProcessedRecordV2 = (
+  record:
+    | IngestJobV2SuccessfulResults<Schema>[number]
+    | IngestJobV2UnprocessedRecords<Schema>[number]
+    | IngestJobV2FailedResults<Schema>[number]
+): BulkProcessedRecordV2 => record as unknown as BulkProcessedRecordV2;
+
+/** call the describe to verify the object exists in the org  */
+export const validateSobjectType = async (sobjectType: string, connection: Connection): Promise<string> => {
   try {
     await connection.sobject(sobjectType).describe();
+    return sobjectType;
   } catch (e) {
     throw new Error(messages.getMessage('invalidSobject', [sobjectType, (e as Error).message]));
   }
 };
-
-export const waitOrTimeout = async (job: IngestJobV2<Schema, IngestOperation>, wait: number): Promise<void> => {
```

Comment:
jsforce now handles the polling/timeout logic and emits events, so this is removed